### PR TITLE
feat:[PLG-2739]: Implement CREATE flow for GitOps Application

### PR DESCRIPTION
**Pull request by neelam-harness**
Added a new sub-command that takes in the YAML to create GitOps Application resource on Harness platform.

Sample YAML:

gitops:
  name: gitops-application
  projectIdentifier: default_project
  orgIdentifier: default
  type: application
  application:
    metadata:
      namespace: default
      clusterName: gitopscluster
      labels:
        harness.io/serviceRef: gitopsservice
        harness.io/envRef: gitopsenv
    spec:
      source:
        repoURL: https://github.com/harness-community/harnesscd-example-apps
        path: guestbook
        targetRevision: master
      destination:
        server: https://kubernetes.default.svc
        namespace: default
  agentIdentifier: AGENT_NAME
  clusterIdentifier: gitopscluster
  repoIdentifier: gitopsrepo
Here, AGENT_NAME is a constant and asks for input from the user.

Command:

harness gitops-application --file applications.yml apply
Asks for AGENT_NAME input from the user:

Trying to create or update gitops-application using the yaml= applications.yml
? Enter a valid AgentIdentifier: debacligitops
Following output:

Not Implemented
Creating GitOps-Application with id:  gitops-application-6
Successfully created GitOps-Application with id= gitops-application-6
Added a new sub-command that takes in the YAML to create GitOps Application resource on Harness platform.

Sample YAML:

gitops:
  name: gitops-application
  projectIdentifier: default_project
  orgIdentifier: default
  type: application
  application:
    metadata:
      namespace: default
      clusterName: gitopscluster
      labels:
        harness.io/serviceRef: gitopsservice
        harness.io/envRef: gitopsenv
    spec:
      source:
        repoURL: https://github.com/harness-community/harnesscd-example-apps
        path: guestbook
        targetRevision: master
      destination:
        server: https://kubernetes.default.svc
        namespace: default
  agentIdentifier: AGENT_NAME
  clusterIdentifier: gitopscluster
  repoIdentifier: gitopsrepo
Here, AGENT_NAME is a constant and asks for input from the user.

Command:

harness gitops-application --file applications.yml apply
Asks for AGENT_NAME input from the user:

Trying to create or update gitops-application using the yaml= applications.yml
? Enter a valid AgentIdentifier: debacligitops
Following output:

Not Implemented
Creating GitOps-Application with id:  gitops-application-6
Successfully created GitOps-Application with id= gitops-application-6

### CODE REVIEW

**Review by neelam-harness — COMMENTED** [review]
LGTM

File: utils.go [hunk 1]
```diff
@@ -269,7 +269,7 @@ func getEntity(baseUrl string, reqURL string, projectIdentifier string, orgIdent
 	queryParams = mergeMaps(queryParams, extraParams)
 	urlWithQueryParams := GetUrlWithQueryParams("", baseUrl, reqURL, queryParams)
 	_, fetchEntityError := Get(urlWithQueryParams, cliCdRequestData.AuthToken)
-
+	fmt.Printf("urltest", urlWithQueryParams)
```
Comment (neelam-harness) [inline utils.go:272]: let's remove this.

File: opcluster.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+gitops:
```
Comment (neelam-harness) [inline opcluster.yml]: do we need this for application ?

**Comment by Debanitrkl** [comment]
Added a new sub-command that takes in the YAML to create GitOps Cluster resource on Harness platform.

Sample YAML:

```
gitops:
  name: gitops-application
  projectIdentifier: default_project
  orgIdentifier: default
  type: application
  application:
    metadata:
      namespace: default
      clusterName: gitopscluster
      labels:
        harness.io/serviceRef: gitopsservice
        harness.io/envRef: gitopsenv
    spec:
      source:
        repoURL: https://github.com/harness-community/harnesscd-example-apps
        path: guestbook
        targetRevision: master
      destination:
        server: https://kubernetes.default.svc
        namespace: default
  agentIdentifier: AGENT_NAME
  clusterIdentifier: gitopscluster
  repoIdentifier: gitopsrepo
 ```
 Here, `AGENT_NAME` is a constant and asks for input from the user.
 
 Command:
 ```
 harness gitops-application --file applications.yml apply
 ```
 
 Asks for AGENT_NAME input from the user:
 
 ```
 Trying to create or update gitops-application using the yaml= applications.yml
? Enter a valid AgentIdentifier: debacligitops
```
Following output:
```
Creating GitOps-Application with id:  gitops-application-2
Data:  {
 "application": {
  "metadata": {
   "name": "gitops-application",
   "namespace": "default",
   "clusterName": "gitopscluster",
   "labels": {
    "harness.io/envRef": "gitopsenv",
    "harness.io/serviceRef": "gitopsservice"
   },
   "Annotations": null
  },
  "spec": {
   "source": {
    "repoURL": "https://github.com/harness-community/harnesscd-example-apps",
    "path": "guestbook",
    "targetRevision": "master"
   },
   "destination": {
    "server": "https://kubernetes.default.svc",
    "namespace": "default"
   }
  }
 }
}
Data:  {}
Successfully created GitOps-Application with id= gitops-application
```

File: client.go [hunk 3]
```diff
@@ -16,7 +17,8 @@ func Post(reqUrl string, auth string, body interface{}, contentType string, requ
 	postBody, _ := json.Marshal(body)
 	requestBody := bytes.NewBuffer(postBody)
 	var req *http.Request
-
+	fmt.Println("requestBody")
```
Comment (neelam-harness) [inline client.go:20]: Please remove these.

File: constants.go [hunk 4]
```diff
@@ -14,6 +14,7 @@ const GITHUB_SECRET_IDENTIFIER = "harness_gitpat"
 const GITOPS_AGENT_IDENTIFIER_PLACEHOLDER = "AGENT_NAME"
 const GITOPS_BASE_URL = "/gateway/gitops/api/v1/agents/"
 const GITOPS_REPOSITORY_ENDPOINT = "repositories"
+const GITOPS_APPLICATION_ENDPOINT = "applications"
```
Comment (neelam-harness) [inline constants.go:17]: Please add it in an alphabetical order.

File: gitops-application.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+import (
+	"fmt"
+	"github.com/fatih/color"
+	"github.com/urfave/cli/v2"
+)
+
+func applyGitopsApplications(c *cli.Context) error {
+	filePath := c.String("file")
+	baseURL := getBaseUrl(c, GITOPS_BASE_URL)
+	// entityURL := strings.TrimSuffix(baseURL, "/agents/")
```
Comment (neelam-harness) [inline gitops-application.go:12]: Please remove this.

File: client.go [hunk 6]
```diff
@@ -16,7 +17,8 @@ func Post(reqUrl string, auth string, body interface{}, contentType string, requ
...
+	printJson(requestBody)
```
Comment (neelam-harness) [inline client.go:21]: Please remove this

File: gitops-application.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+	createOrUpdateApplicationURL := GetUrlWithQueryParams("", baseURL, GITOPS_APPLICATION_ENDPOINT, map[string]string{
+		"accountIdentifier": cliCdRequestData.Account,
+		"orgIdentifier":     orgIdentifier,
+		"projectIdentifier": projectIdentifier,
+		"clusterIdentifier": clusterIdentifier,
+		"repoIdentifier":    repoIdentifier,
+	})
+	extraParams := map[string]string{
+
```
Comment (neelam-harness) [inline gitops-application.go:43]: Nit: remove empty line.

File: gitops-application.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+			println(getColoredText("Successfully created GitOps-Application with id= ", color.FgGreen) +
+				getColoredText(applicationName, color.FgBlue))
+			return nil
+		}
+	} else {
+		println("Found GitOps-Application with id=", getColoredText(applicationName, color.FgCyan))
+		println("Updating details of GitOps-Application with id=", getColoredText(applicationName, color.FgBlue))
+
+		var appPUTUrl = GetUrlWithQueryParams("", baseURL,
+			//fmt.Sprintf("%s/%s/spec", GITOPS_APPLICATION_ENDPOINT, applicationName), map[string]string{
```
Comment (neelam-harness) [inline gitops-application.go:68]: Please remove the commented code.

File: client.go [hunk 9]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/json"
 	"errors"
+	_ "fmt"
```
Comment (neelam-harness) [inline client.go:7]: do we need this ?

File: gitops-application.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+	entityExists := getEntity(baseURL, fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT+"/%s"+"/exists", applicationName),
+		projectIdentifier, orgIdentifier, extraParams)
+	var _ ResponseBody
+	var err error
+	if !entityExists {
+		println("Creating GitOps-Application with id: ", getColoredText(applicationName, color.FgGreen))
+		applicationPayload := createGitOpsApplicationPayload(requestBody)
+
+		printJson(applicationPayload)
```
Comment (neelam-harness) [inline gitops-application.go:55]: We need to remove this.

File: gitops-cluster.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
```
Comment (neelam-harness) [inline gitops-cluster.go:3]: This file is already checked in, please make sure this doesn't override the existing file. Seems like a rebase issue.

File: types.go [hunk 12]
```diff
@@ -251,20 +280,48 @@ type Repo struct {
...
-	Paths []string `json:"Paths"`
-}
-
 type RepoWithUpdateMask struct {
 	Repo       `json:"repo"`
 	UpdateMask struct {
 		Paths []string `json:"paths"`
 	} `json:"updateMask"`
 }
+type UpdateMask struct {
```
Comment (neelam-harness) [inline types.go:289]: I don't think `UpdateMask` is required anymore, since it's being declared internally in line# 285.

File: utils.go [hunk 13]
```diff
@@ -97,7 +97,6 @@ func readFromFile(filepath string) (s string, r []byte) {
 		//log.Println("Error reading from file:", fileError)
 	}
 	_fileContents = string(byteValue)
-
```
Comment (neelam-harness) [inline utils.go:100]: Let's not make any changes in the file that is not being touched by the current use-case.

File: constants.go [hunk 14]
```diff
@@ -11,6 +11,7 @@ const DEFAULT_PROJECT = "default_project"
 const DEFAULT_ORG = "default"
 const ENVIRONMENT_ENDPOINT = "environmentsV2"
 const GITHUB_SECRET_IDENTIFIER = "harness_gitpat"
+const GITOPS_APPLICATION_ENDPOINT = "applications"
 const GITOPS_AGENT_IDENTIFIER_PLACEHOLDER = "AGENT_NAME"
```
Comment (Ompragash) [inline constants.go:15]:
```suggestion
const GITOPS_AGENT_IDENTIFIER_PLACEHOLDER = "AGENT_NAME"
const GITOPS_APPLICATION_ENDPOINT = "applications"
```

**Review by Ompragash — CHANGES_REQUESTED** [review]
Reordered constants to preserve the chronological sequence.

**Review by Ompragash — APPROVED** [review]
Tested the changes by creating GitOps resources on Harness and LGTM!
Thanks @Debanitrkl!

**Pull request merged**